File: uva10344/uva10344.py
from sys import stdin



def make_perms(n):
    global path,perms,visited
    if(n == 5):
        perms.append(list(path))
    else:
        for i in range(5):
            if not visited[i]:
                path.append(i)
                visited[i] = True
                make_perms(n+1)
                visited[i] = False
                path.pop()
        

def solve(s,N):
    ans = s == 23 and N == 5
    if N < 5: ans = solve(s+nums[N],N+1) or solve(s-nums[N],N+1) if s*nums[N] > 200 else solve(s+nums[N],N+1) or solve(s-nums[N],N+1) or solve(s*nums[N],N+1)
    return ans

def main():
    global path,perms,visited,nums
    perms,path,visited,nums,i,ans,j,inp = list(),list(),[False for _ in range(5)],list(),0,None,0,None
    make_perms(0)
    inp = [int(x) for x in stdin.readline().strip().split()]
    while not(inp[0] == 0):
        i,ans = 0,False
        while i < len(perms) and not ans:
            nums = list()
            for j in perms[i]: nums.append(inp[j])
            ans,i = ans or solve(nums[0],1),i+1
        print("Impossible" if not ans else "Possible")
        inp = [int(x) for x in stdin.readline().strip().split()]
                
            
        

main()

File: uva10344/uva10344.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <algorithm>
using namespace std;

int numbers[5];



int solve(int N,int v){
	int ans;
	if(N == 5 && v != 23)
		ans = 0;
	else if(N == 5 && v == 23)
		ans = 1;
	else{
		ans = solve(N+1,v+numbers[N]) || solve(N+1,v-numbers[N]) || solve(N+1,v*numbers[N]);
	}

	return ans;
}

int main(){
	int i,ans;
	for(i = 0; i < 5;i++)
		scanf("%d",&numbers[i]);
	while(!(numbers[0] == 0 && numbers[1] == 0 && numbers[2] == 0 && numbers[3] == 0 && numbers[4] == 0)){
		sort(numbers,numbers+5);
		ans = 0;
		do{
			ans = ans || solve(1,numbers[0]);
		}while(next_permutation(numbers,numbers+5) && !ans);
		if(ans)
			printf("Possible\n");
		else
			printf("Impossible\n");


		for(i = 0; i < 5;i++)
			scanf("%d",&numbers[i]);


	}
	return 0;
}